src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs

```diff
@@ -0,0 +1,145 @@
+using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using RoslyRenamer = Microsoft.CodeAnalysis.Rename;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    internal class Renamer
+    {
+        private readonly VisualStudioWorkspace _visualStudioWorkspace;
+        private readonly IProjectThreadingService _threadingService;
+        private readonly SVsServiceProvider _serviceProvider;
+        private readonly Project _project;
+        private readonly Document _oldDocument;
+        private readonly string _newFilePath;
+        private bool _docAdded = false;
+        private bool _docRemoved = false;
+        private bool _docChanged = false;
+
+        internal Renamer(VisualStudioWorkspace visualStudioWorkspace,
+                         SVsServiceProvider serviceProvider,
+                         IProjectThreadingService threadingService,
+                         Project project,
+                         string newFilePath,
+                         string oldFilePath)
+        {
+            _visualStudioWorkspace = visualStudioWorkspace;
+            _serviceProvider = serviceProvider;
+            _threadingService = threadingService;
+            _project = project;
+            _newFilePath = newFilePath;
+            _oldDocument = (from d in project.Documents where d.FilePath.Equals(oldFilePath, StringComparison.OrdinalIgnoreCase) select d).FirstOrDefault();
+        }
+
+        public void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs args)
+        {
+            if (args.Kind == WorkspaceChangeKind.DocumentAdded && args.ProjectId == _project.Id)
+            {
+                Project project = (from p in args.NewSolution.Projects where p.Id.Equals(_project.Id) select p).FirstOrDefault();
+                Document addedDocument = (from d in project.Documents where d.Id.Equals(args.DocumentId) select d).FirstOrDefault();
+                if (addedDocument.FilePath.Equals(_newFilePath, StringComparison.OrdinalIgnoreCase))
+                {
+                    _docAdded = true;
+                }
+            }
+
+            if (args.Kind == WorkspaceChangeKind.DocumentRemoved && args.ProjectId == _project.Id && args.DocumentId == _oldDocument.Id)
+            {
+                _docRemoved = true;
+            }
+
+            if (args.Kind == WorkspaceChangeKind.DocumentChanged && args.ProjectId == _project.Id)
+            {
+                _docChanged = true;
+            }
+
+            if (_docAdded && _docRemoved && _docChanged)
+            {
+                _visualStudioWorkspace.WorkspaceChanged -= OnWorkspaceChanged;
+
+                _threadingService.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    var myNewProject = _visualStudioWorkspace.CurrentSolution.Projects.Where(p => string.Equals(p.FilePath, _project.FilePath, StringComparison.OrdinalIgnoreCase)).FirstOrDefault();
```

Comment:
You're reading some things here, such as project.FilePath on a background thread.